src/ucs/sys/topo/base/topo.c

```diff
@@ -324,6 +324,49 @@ const char *ucs_topo_distance_str(const ucs_sys_dev_distance_t *distance,
     return ucs_string_buffer_cstr(&strb);
 }
 
+ucs_sys_device_t ucs_topo_set_sys_dev(const char *dev_name,
+                                      const char *sysfs_path,
+                                      int is_primary)
+{
+    const char *bdf_name;
+    ucs_status_t status;
+    ucs_sys_device_t sys_dev;
+    int num_devices_pre, override_name;
+
+    if (sysfs_path == NULL) {
+        goto out_unknown;
+    }
+
+    bdf_name = strrchr(sysfs_path, '/');
+    if (bdf_name == NULL) {
+        goto out_unknown;
+    }
+
+    ++bdf_name; /* Move past '/' separator */
+
+    /* Store number of devices before we add our device */
+    num_devices_pre = ucs_topo_num_devices();
+
+    status = ucs_topo_find_device_by_bdf_name(bdf_name, &sys_dev);
+    if (status != UCS_OK) {
+        goto out_unknown;
+    }
+
+    /* Override device name if device is new (number of devices increased),
+       or if it's a primary device */
+    override_name = (num_devices_pre < ucs_topo_num_devices()) || is_primary;
+    status = ucs_topo_sys_device_set_name(sys_dev, dev_name, override_name);
+
+    ucs_assert_always(status == UCS_OK);
```

Comment:
if failed return unknown

Reply:
won't it cause a corruption in `ucs_topo_global_ctx.devices` ?

Reply:
it's created with a default name.. so now i understand why passing override_name directly is not good enough
maybe we should keep a field in ucs_topo_global_ctx.devices (per device) that means the "priority" of the name.. and compare it with a priority argument provided to ucs_topo_sys_device_set_name()